4-Firebase-Firestore/end/src/App.js

```diff
@@ -0,0 +1,384 @@
+import LoginForm from "./components/LoginForm";
+import AddEditRecipeForm from "./components/AddEditRecipeForm";
+import "./App.css";
+
+function App() {
+  const [user, setUser] = useState(null);
+  const [currentRecipe, setCurrentRecipe] = useState(null);
+  const [recipes, setRecipes] = useState([]);
+  const [isLoading, setIsloading] = useState(false);
+  const [categoryFilter, setCategoryFilter] = useState("");
+  const [orderBy, setOrderBy] = useState("unixTimeStampInMillisecondsDesc");
+  const [recipesPerPage, setRecipesPerPage] = useState(3);
+  const [recipesCount, setRecipesCount] = useState(null);
+
+  useEffect(() => {
+    setIsloading(true);
+    fetchRecipes()
+      .then((fetchedRecipes) => {
+        setRecipes(fetchedRecipes);
+      })
+      .catch((error) => {
+        console.error(error.message);
+        throw error;
+      })
+      .finally(() => {
+        setIsloading(false);
+      });
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [user, categoryFilter, orderBy, recipesPerPage]);
+
+  FirebaseAuthService.subscribeToAuthChanges(setUser);
+
+  async function fetchRecipes(cursorId = "") {
+    const queries = [];
+    if (categoryFilter) {
+      queries.push({
+        field: "category",
+        condition: "==",
+        value: categoryFilter,
+      });
+    }
+    if (!user) {
+      queries.push({
+        field: "isPublished",
+        condition: "==",
+        value: true,
+      });
+    }
+
+    const orderByField = "unixTimeStampInMilliseconds";
+    let orderByDirection;
+
+    if (orderBy) {
+      switch (orderBy) {
+        case "unixTimeStampInMillisecondsAsc":
+          orderByDirection = "asc";
+          break;
+        case "unixTimeStampInMillisecondsDesc":
+          orderByDirection = "desc";
+          break;
+        default:
+          break;
+      }
+    }
+    try {
+      const response = await FirebaseFireStoreService.readDocuments({
+        collection: "recipes",
+        queries: queries,
+        orderByField: orderByField,
+        orderByDirection: orderByDirection,
+        perPage: recipesPerPage,
+        cursorId: cursorId,
+      });
+
+      let newRecipes = response.results.docs.map((doc) => ({
+        id: doc.id,
+        ...doc.data(),
+      }));
+
+      setRecipesCount(response.recipeSize);
+
+      if (cursorId) {
+        let fetchedRecipes = [...recipes, ...newRecipes];
+
+        return fetchedRecipes;
+      } else {
+        let fetchedRecipes = response.results.docs.map((doc) => ({
+          id: doc.id,
+          ...doc.data(),
+        }));
+        return fetchedRecipes;
+      }
+    } catch (error) {
+      console.error(error.message);
+      throw error;
+    }
+  }
+
+  function handleRecipesPerPageChange(event) {
+    const recipesPerPage = event.target.value;
+
+    startTransition(() => {
+      setRecipes([]);
+      setRecipesPerPage(recipesPerPage);
+    });
+  }
```

Comment:
Please add a new line so these functions aren't butted up right against each other. It makes it harder to read.